File: api/internal/crdt/fracindex.go
package crdt

// Fractional indexing generates position strings that sort lexicographically.
// Used for task and subtask ordering without reindexing.
//
// Uses the full printable ASCII range (! through ~) so positions work with
// any string format (digits, letters, mixed).

const minChar = '!' // 33
const maxChar = '~' // 126
const midChar = 'O' // 79

// First returns a position before all existing items.
func First() string {
	return string(midChar)
}

// Last returns a position after all existing items.
func Last() string {
	return string(maxChar)
}

// Between generates a position string that sorts between before and after.
// If before is empty, generates a position before after.
// If after is empty, generates a position after before.
func Between(before, after string) string {
	if before == "" {
		before = string(minChar)
	}
	if after == "" {
		after = string(maxChar)
	}

	// Ensure before < after
	if before >= after {
		return before + string(midChar)
	}

	// Pad shorter string with minChar to equal length
	maxLen := len(before)
	if len(after) > maxLen {
		maxLen = len(after)
	}

	bPadded := padRight(before, maxLen)
	aPadded := padRight(after, maxLen)

	// Find midpoint character by character
	for i := 0; i < maxLen; i++ {
		bChar := bPadded[i]
		aChar := aPadded[i]

		if bChar < aChar {
			mid := bChar + (aChar-bChar)/2
			if mid > bChar {
				return before[:min(i, len(before))] + string(rune(mid))
			}
		}
	}

	// No room between — append a middle character after `before`
	return before + string(midChar)
}

func padRight(s string, length int) []byte {
	padded := make([]byte, length)
	copy(padded, s)
	for i := len(s); i < length; i++ {
		padded[i] = minChar
	}
	return padded
}

File: web/src/crdt/fracindex.ts
// Printable ASCII range for position strings.
// Supports any mix of digits, letters, and symbols.
const MIN_CHAR = '!'.charCodeAt(0) // 33
const MAX_CHAR = '~'.charCodeAt(0) // 126
const MID_CHAR = 'O'.charCodeAt(0) // 79 — middle of printable range

export function first(): string {
  return String.fromCharCode(MID_CHAR)
}

export function last(): string {
  return String.fromCharCode(MAX_CHAR)
}

/**
 * Generate a position string that sorts lexicographically between `before` and `after`.
 * If before is empty, generates a position before after.
 * If after is empty, generates a position after before.
 * Works with any printable ASCII strings (digits, letters, mixed).
 */
export function between(before: string, after: string): string {
  if (!before) before = String.fromCharCode(MIN_CHAR)
  if (!after) after = String.fromCharCode(MAX_CHAR)

  // Ensure before < after lexicographically
  if (before >= after) {
    return before + String.fromCharCode(MID_CHAR)
  }

  const maxLen = Math.max(before.length, after.length)
  const bPadded = padRight(before, maxLen)
  const aPadded = padRight(after, maxLen)

  for (let i = 0; i < maxLen; i++) {
    const bChar = bPadded.charCodeAt(i)
    const aChar = aPadded.charCodeAt(i)

    if (bChar < aChar) {
      const mid = bChar + Math.floor((aChar - bChar) / 2)
      if (mid > bChar) {
        return before.slice(0, Math.min(i, before.length)) + String.fromCharCode(mid)
      }
    }
  }

  // No room between — append a middle character after `before`
  return before + String.fromCharCode(MID_CHAR)
}

function padRight(s: string, length: number): string {
  while (s.length < length) s += String.fromCharCode(MIN_CHAR)
  return s
}
